server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,650 @@
+            " use --config <file> on the command line".format(_NAME_),
+            file=sys.stderr,
+        )
+        return 2
+
+    try:
+        config = PbenchServerConfig(cfg_name)
+    except BadConfig as e:
+        print("{}: {} (config file {})".format(_NAME_, e, cfg_name), file=sys.stderr)
+        return 1
+
+    if check_func(config.ARCHIVE, "ARCHIVE") > 0:
+        return 1
+
+    if check_func(config.INCOMING, "INCOMING") > 0:
+        return 1
+
+    if check_func(config.RESULTS, "RESULTS") > 0:
+        return 1
+
+    if check_func(config.USERS, "USERS") > 0:
+        return 1
+
+    logger = get_pbench_logger(_NAME_, config)
+
+    ret = 0
+
+    try:
+        os.mkdir(os.path.join(config.LOGSDIR, "pbench-audit-server"))
+    except FileExistsError:
+        # directory already exists, ignore
+        pass
+    except Exception:
+        print("os.mkdir: Unable to create destination directory")
+        return 1
+    logfile = Path(config.LOGSDIR, "pbench-audit-server", "pbench-audit-server.log")
+    # TEMPORARY addition of error file for the sake of test cases
+    errorfile = Path(config.LOGSDIR, "pbench-audit-server", "pbench-audit-server.error")
+    with errorfile.open(mode="w") as f:
+        pass
+    # END
+
+    logger.info("start-{}", config.TS)
+
+    with logfile.open(mode="w") as f:
+
+        ahier = ArchiveHierarchy("archive", config.ARCHIVE, config)
+        verify_archive(ahier)
+        cnt = 0
+        if ahier.check_controller_in_archive():
+            f.write(f"\nstart-{config.TS[4:]}: archive hierarchy: {config.ARCHIVE}\n")
+            cnt = ahier.dump(f)
+            ahier.header(f, "end")
+        if cnt > 0:
+            ret += 1
+
+        ihier = IncomingHierarchy("incoming", config.INCOMING, config)
+        cihier = ControllerHierarchy("incoming", config.INCOMING, config)
+        verify_controllers(ihier, cihier)
+        cnt = 0
+        c_check = cihier.check_controller_in_controller()
+        i_check = ihier.check_controller_in_IRHierarchy()
+        if c_check or i_check:
+            cihier.header(f, "start")
+            if c_check:
+                cnt = cihier.dump(f)
+            if i_check:
+                cnt += ihier.dump(f)
+            cihier.header(f, "end")
+        if cnt > 0:
+            ret += 1
+
+        rhier = ResultsHierarchy("results", config.RESULTS, config)
+        crhier = ControllerHierarchy("results", config.RESULTS, config)
+        verify_controllers(rhier, crhier)
+        cnt = 0
+        c_check = crhier.check_controller_in_controller()
+        r_check = rhier.check_controller_in_IRHierarchy()
+        if c_check or r_check:
+            crhier.header(f, "start")
+            if c_check:
+                cnt = crhier.dump(f)
+            if r_check:
+                cnt += rhier.dump(f)
+            crhier.header(f, "end")
+        if cnt > 0:
+            ret += 1
+
+        ruhier = ResultsHierarchy("results", config.RESULTS, config)
```

Comment:
Is there any reason we can't reuse the ResultsHierarchy and ControllerHierarchy builds from lines 595 and 596? I get that verify_users() does a different level of scan with the optional hierarchy parameter to verify_controllers, but it's hard to understand the detailed expected mutations of the objects based on that. If they're really not reusable (or can't be easily cleaned), then a comment explaining that might be helpful.

Reply:
Sure, I will add a comment explaining that

Reply:
This is not correct: there isn't a single ResultsHierarchy here: the Users hierarchy consists of a sequence of Results hierarchies, one per user - we need to get the list of directories under the .../users path (and mark non-directories as bad) and for each of those directories (i.e. for each user), we need to construct a Results hierarchy (the controllers, and the links under each of those controllers).

That will entail changes to the User hierarchy implementation and the way that the audit server calls into it, but it will allow simplifications: `verify_controllers()` does not need the `user_arg` (and the `verify_incoming()` and `verify_results()` functions can lose it as well), plus there is no need for the dispatching function `verify_hierarchy()` then: we can delete the call from `verify_controllers()` and call either `verify_incoming()` or `verify_results()` explicitly in `main()`; plus each one can take a single hierarchy as argument: no more tangling together of multiple hierarchies in each function.

@riay-17 and I talked earlier today about this: she'll implement the simplifications first and then rework the User hierarchy along the lines in the first paragraph above.